File: src/lib/events.ts
/**
 * Typed Tauri event listeners.
 *
 * Mirrors the `#[serde(tag = "type", rename_all = "snake_case")]` enums
 * defined in `src-tauri/src/events.rs`.  The `type` discriminant is always
 * present; all other fields are optional to accommodate every variant.
 */

import { listen, type UnlistenFn } from '@tauri-apps/api/event'
import type { UUID } from './types'
import type { NodeDocument } from '@/lib/bindings'

// ============================================================================
// Agent events  ("agent-event")
// ============================================================================

export type AgentEventType =
  | 'started'
  | 'token'
  | 'tool_call'
  | 'tool_result'
  | 'done'
  | 'error'
  | 'persisted'
  | 'tool_approval_required'
  | 'stream_update'
  | 'thinking_stream_update'
  | 'thinking_done'

export interface AgentEvent {
  type: AgentEventType
  conversation_id: string
  delta?: string
  tool_call?: ToolCallInfo
  tool_call_id?: string
  result?: string
  input_tokens?: number
  output_tokens?: number
  message?: string
  tool_name?: string
  arguments?: Record<string, unknown>
  draft_html?: string | null
  slot_count?: number
  new_toc_items?: any[]
  new_artifact_specs?: any[]
  thinking_document?: NodeDocument
}
export interface ToolCallInfo {
  id: string
  name: string
  arguments: Record<string, unknown>
}

// ============================================================================
// MCP events  ("mcp-event")
// ============================================================================

export type McpEventType =
  | 'server_connected'
  | 'server_disconnected'
  | 'tool_discovered'

export interface McpEvent {
  type: McpEventType
  name?: string
  server?: string
  tool?: McpToolInfo
}

export interface McpToolInfo {
  name: string
  description: string
}

// ============================================================================
// Workflow events  ("workflow-event")
// ============================================================================

export type WorkflowEventType =
  | 'started'
  | 'step_started'
  | 'step_completed'
  | 'completed'
  | 'failed'

export interface WorkflowEvent {
  type: WorkflowEventType
  id?: UUID
  workflow_id?: UUID
  step_id?: string
  result?: string
  message?: string
}

// ============================================================================
// Skill events  ("skill-event")
// ============================================================================

export type SkillEventType = 'updated'

export interface SkillEvent {
  type: SkillEventType
  source_label: string
  skill_name: string
}

// ============================================================================
// Queue events  ("queue-event") – NEW
// ============================================================================

export type QueueEventType = 'message_sent'

export interface QueueEvent {
  type: QueueEventType
  conversation_id: string
  message_id: string
}

// ============================================================================
// Listener helpers
// ============================================================================

export function onAgentEvent(
  callback: (event: AgentEvent) => void
): Promise<UnlistenFn> {
  return listen<AgentEvent>('agent-event', (e) => callback(e.payload))
}

export function onMcpEvent(
  callback: (event: McpEvent) => void
): Promise<UnlistenFn> {
  return listen<McpEvent>('mcp-event', (e) => callback(e.payload))
}

export function onWorkflowEvent(
  callback: (event: WorkflowEvent) => void
): Promise<UnlistenFn> {
  return listen<WorkflowEvent>('workflow-event', (e) => callback(e.payload))
}

export function onSkillEvent(
  callback: (event: SkillEvent) => void
): Promise<UnlistenFn> {
  return listen<SkillEvent>('skill-event', (e) => callback(e.payload))
}

// NEW: Queue event listener
export function onQueueEvent(
  callback: (event: QueueEvent) => void
): Promise<UnlistenFn> {
  return listen<QueueEvent>('queue-event', (e) => callback(e.payload))
}
export type RunCodeEventType = 'stdout' | 'stderr' | 'exit'

export interface RunCodeEvent {
  type: RunCodeEventType
  run_id: string
  line: string
  code?: number
  elapsed_ms?: number
}

File: src-tauri/src/events.rs
//! Typed IPC event payloads emitted from Rust to the React frontend.
//!
//! All event names use kebab-case and match the listeners in `src/lib/events.ts`.

use serde::{Deserialize, Serialize};
use skilldeck_core::markdown::types::{ArtifactSpec, NodeDocument, TocItem};
use specta::Type;
use tauri_specta::Event;

// ── Agent events ─────────────────────────────────────────────────────────────

/// Payload for the `"agent-event"` Tauri channel.
#[derive(Debug, Clone, Serialize, Deserialize, Type, Event)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum AgentEvent {
    Cancelled {
        conversation_id: String,
    },
    Started {
        conversation_id: String,
    },
    Token {
        conversation_id: String,
        delta: String,
    },
    ToolCall {
        conversation_id: String,
        tool_call: AgentToolCall,
    },
    ToolResult {
        conversation_id: String,
        tool_call_id: String,
        result: String,
    },
    Done {
        conversation_id: String,
        input_tokens: u32,
        output_tokens: u32,
    },
    Error {
        conversation_id: String,
        message: String,
    },
    Persisted {
        conversation_id: String,
    },
    ToolApprovalRequired {
        conversation_id: String,
        tool_call_id: String,
        tool_name: String,
        arguments: serde_json::Value,
    },
    StreamUpdate {
        conversation_id: String,
        document: NodeDocument,
        new_toc_items: Vec<TocItem>,
        new_artifact_specs: Vec<ArtifactSpec>,
    },
    ProviderNotReady {
        conversation_id: String,
        reason: String,
        fix_action: String,
    },
    ThinkingStreamUpdate {
        conversation_id: String,
        document: NodeDocument,
    },
    ThinkingDone {
        conversation_id: String,
        document: NodeDocument,
    },
}

#[derive(Debug, Clone, Serialize, Deserialize, Type)]
pub struct AgentToolCall {
    pub id: String,
    pub name: String,
    pub arguments: serde_json::Value,
}

// ── MCP events ────────────────────────────────────────────────────────────────

/// Payload for the `"mcp-event"` Tauri channel.
#[derive(Debug, Clone, Serialize, Deserialize, Type, Event)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum McpEvent {
    ServerConnected { name: String },
    ServerDisconnected { name: String },
    ServerFailed { name: String, message: String },
    ToolDiscovered { server: String, tool: McpToolInfo },
}

#[derive(Debug, Clone, Serialize, Deserialize, Type)]
pub struct McpToolInfo {
    pub name: String,
    pub description: String,
}

// ── Workflow events ───────────────────────────────────────────────────────────

/// Payload for the `"workflow-event"` Tauri channel.
#[derive(Debug, Clone, Serialize, Deserialize, Type, Event)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum WorkflowEvent {
    Started {
        id: String,
    },
    StepStarted {
        workflow_id: String,
        step_id: String,
    },
    StepCompleted {
        workflow_id: String,
        step_id: String,
        result: Option<String>,
    },
    StepFailed {
        workflow_id: String,
        step_id: String,
        error: String,
    },
    Completed {
        id: String,
    },
    Failed {
        id: String,
        message: String,
    },
}

// ── Skill events ─────────────────────────────────────────────────────────────

/// Payload for the `"skill-event"` Tauri channel.
#[derive(Debug, Clone, Serialize, Deserialize, Type, Event)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum SkillEvent {
    Updated {
        source_label: String,
        skill_name: String,
    },
}

// ── Queue events ────────────────────────────────────────────────────────────
/// Payload for the `"queue-event"` Tauri channel.
#[derive(Debug, Clone, Serialize, Deserialize, Type, Event)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum QueueEvent {
    MessageSent {
        conversation_id: String,
        message_id: String,
    },
}

#[derive(Debug, Clone, Serialize, Deserialize, Type, Event)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum RunCodeEvent {
    Stdout {
        run_id: String,
        line: String,
    },
    Stderr {
        run_id: String,
        line: String,
    },
    Exit {
        run_id: String,
        code: i32,
        elapsed_ms: u64,
    },
}
